src/openclsim/plot/superlog.py

```diff
@@ -0,0 +1,459 @@
+        """dummy docstring"""
+        assert set(self.COLUMNS_SUPERLOG).issubset(
+            set(df_super_log.columns)
+        ), "df_super_log input is missing columns"
+
+        # init
+        # pd.Dataframe with columns as defince and optionally column 'is_critical' (and others)
+        self.df_super_log = df_super_log
+        # list of tuples e.g. [(A, C), (B, C), (C, F)]  when C depends on A and B and F on C
+        self.dependencies = self._get_dependencies()
+
+    def _get_dependencies(self):
+        """
+        set dependencies (list of tuples with cp_activity_id vales).
+        Find dependencies within all cp_activities based on timestamps.
+        Allow dependencies with unequal cp_activities only when waiting.
+        In this special case the code assume dependency when an activity stops
+        and another activity starts at the exact same time
+
+        Returns
+        ------------
+        dependencies : list
+            list of tuplese e.g. [(A, C), (B, C), (C, F)]  when C depends on A and B and F on C
+        """
+        list_dependencies = []
+        # loop over each unique cp_activity
+        for cp_act in self.df_super_log.loc[:, "cp_activity_id"].unique():
+            # and find cp_activities that END when this one STARTS
+            bool_this_activity = self.df_super_log.loc[:, "cp_activity_id"] == cp_act
+            bool_end_when_start = (
+                self.df_super_log.loc[:, "end_time"]
+                == self.df_super_log.loc[bool_this_activity, "start_time"].iloc[0]
+            )
+            bool_shared_source = self.df_super_log.loc[:, "SourceObject"].isin(
+                list(self.df_super_log.loc[bool_this_activity, "SourceObject"])
+            )
+            # so standard dependencies requires identical time and (at least 1) shared Source Object
+            bool_dependencies = (
+                bool_shared_source & bool_end_when_start & ~bool_this_activity
+            )
+            if sum(bool_dependencies) > 0:
+                dependencies = self.df_super_log.loc[bool_dependencies, :]
+                if "WAITING" in dependencies.loc[:, "state"].tolist():
+                    # this activity depends on waiting meaning
+                    # current activity might have been waiting on activity of other source object!
+                    # we ASSUME that every activity of every source object with identical end time of this WAITING is a
+                    # dependency for current activity
+                    dependencies = self.df_super_log.loc[
+                        bool_end_when_start & ~bool_this_activity, :
+                    ]
+                # get all unique activities on which current cp_act depends
+                activities_depending = dependencies.loc[:, "cp_activity_id"].unique()
+                for act_dep in activities_depending:
+                    list_dependencies.append((act_dep, cp_act))
+
+        return list_dependencies
```

Comment:
As I understand this is the core of the entire critical path analysis right?

Looks very nice!